File: pastel_palette.php
<?php
// pastel_palette.php — PHP версия

class PastelPalette {
    private $count;
    private $seed;
    private $colors = [];
    private $rng;

    public function __construct($count = 5, $seed = null) {
        $this->count = $count;
        $this->seed = $seed;
        if ($seed !== null) {
            mt_srand($seed);
        }
        $this->rng = function() {
            return mt_rand() / mt_getrandmax();
        };
    }

    private function hslToHex($h, $s, $l) {
        // Конвертация HSL в HEX
        $h = $h / 360;
        $s = $s / 100;
        $l = $l / 100;
        list($r, $g, $b) = $this->hslToRgb($h, $s, $l);
        return sprintf("#%02x%02x%02x", round($r * 255), round($g * 255), round($b * 255));
    }

    private function hslToRgb($h, $s, $l) {
        if ($s == 0) {
            return [$l, $l, $l];
        }
        $c = (1 - abs(2 * $l - 1)) * $s;
        $x = $c * (1 - abs(fmod($h / (1/6), 2) - 1));
        $m = $l - $c / 2;
        if ($h < 1/6) {
            list($r, $g, $b) = [$c, $x, 0];
        } elseif ($h < 2/6) {
            list($r, $g, $b) = [$x, $c, 0];
        } elseif ($h < 3/6) {
            list($r, $g, $b) = [0, $c, $x];
        } elseif ($h < 4/6) {
            list($r, $g, $b) = [0, $x, $c];
        } elseif ($h < 5/6) {
            list($r, $g, $b) = [$x, 0, $c];
        } else {
            list($r, $g, $b) = [$c, 0, $x];
        }
        return [$r + $m, $g + $m, $b + $m];
    }

    public function generate() {
        $this->colors = [];
        for ($i = 0; $i < $this->count; $i++) {
            $h = mt_rand(0, 359);
            $s = mt_rand(20, 40);
            $l = mt_rand(75, 95);
            $hex = $this->hslToHex($h, $s, $l);
            $this->colors[] = [
                'hsl' => "hsl($h, {$s}%, {$l}%)",
                'hex' => $hex,
                'rgb' => "rgb(" . round($h/360*255) . ", " . round($s/100*255) . ", " . round($l/100*255) . ")"
            ];
        }
        return $this->colors;
    }

    public function printPalette() {
        echo "\n📊 Палитра из " . count($this->colors) . " цветов" . ($this->seed !== null ? " (seed: {$this->seed})" : "") . "\n\n";

        // Цветные блоки
        foreach ($this->colors as $c) {
            $hex = substr($c['hex'], 1);
            $r = hexdec(substr($hex, 0, 2));
            $g = hexdec(substr($hex, 2, 2));
            $b = hexdec(substr($hex, 4, 2));
            echo "\033[48;2;{$r};{$g};{$b}m██████████\033[0m ";
        }
        echo "\n";
        foreach ($this->colors as $c) {
            printf("  %-8s", $c['hex']);
        }
        echo "\n";
    }

    public function saveJSON($filename = 'palette.json') {
        $data = ['seed' => $this->seed, 'colors' => $this->colors];
        file_put_contents($filename, json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
        echo "💾 Сохранено JSON: $filename\n";
    }

    public function saveHTML($filename = 'palette.html') {
        $html = '<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Pastel Palette</title>
<style>
body { font-family: monospace; background: #f5f5f5; padding: 20px; }
.palette { display: flex; gap: 20px; justify-content: center; margin-top: 30px; flex-wrap: wrap; }
.color { width: 120px; height: 150px; border-radius: 12px; display: flex; flex-direction: column; align-items: center; justify-content: flex-end; padding: 10px; color: #333; font-weight: bold; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }
.hex { background: rgba(255,255,255,0.7); padding: 4px 8px; border-radius: 4px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🎨 Pastel Palette</h1>
<div class="palette">
';
        foreach ($this->colors as $c) {
            $html .= "<div class=\"color\" style=\"background: {$c['hex']};\">";
            $html .= "<span class=\"hex\">{$c['hex']}</span>";
            $html .= "<span style=\"font-size:12px;\">{$c['hsl']}</span>";
            $html .= "</div>";
        }
        $html .= '
</div>
</body>
</html>';
        file_put_contents($filename, $html);
        echo "💾 Сохранено HTML: $filename\n";
    }
}

function main($argv) {
    $count = 5;
    $seed = null;
    $jsonFile = 'palette.json';
    $htmlFile = 'palette.html';

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--count' || $argv[$i] == '-c') {
            $count = (int)$argv[++$i];
        } elseif ($argv[$i] == '--seed' || $argv[$i] == '-s') {
            $seed = (int)$argv[++$i];
        } elseif ($argv[$i] == '--output-json' || $argv[$i] == '-j') {
            $jsonFile = $argv[++$i];
        } elseif ($argv[$i] == '--output-html' || $argv[$i] == '-h') {
            $htmlFile = $argv[++$i];
        }
    }

    echo "🎨 Pastel Palette Generator (PHP)\n";
    $gen = new PastelPalette($count, $seed);
    $gen->generate();
    $gen->printPalette();
    $gen->saveJSON($jsonFile);
    $gen->saveHTML($htmlFile);
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>

File: pastel_palette.rs
// pastel_palette.rs — Rust версия

use rand::Rng;
use rand::SeedableRng;
use rand::rngs::StdRng;
use std::collections::HashMap;
use std::fs;
use std::env;

struct PastelPalette {
    count: usize,
    seed: u64,
    colors: Vec<HashMap<String, String>>,
    rng: StdRng,
}

impl PastelPalette {
    fn new(count: usize, seed: u64) -> Self {
        let rng = if seed != 0 {
            StdRng::seed_from_u64(seed)
        } else {
            StdRng::from_entropy()
        };
        PastelPalette {
            count,
            seed,
            colors: Vec::new(),
            rng,
        }
    }

    fn hsl_to_hex(h: f64, s: f64, l: f64) -> String {
        // Конвертация HSL в HEX
        let (r, g, b) = hsl_to_rgb(h, s, l);
        format!("#{:02x}{:02x}{:02x}", (r * 255.0) as u8, (g * 255.0) as u8, (b * 255.0) as u8)
    }

    fn generate(&mut self) {
        self.colors.clear();
        for _ in 0..self.count {
            let h = self.rng.gen_range(0.0..360.0);
            let s = self.rng.gen_range(20.0..40.0);
            let l = self.rng.gen_range(75.0..95.0);
            let hex = Self::hsl_to_hex(h, s, l);
            let mut color = HashMap::new();
            color.insert("hsl".to_string(), format!("hsl({:.0}, {:.0}%, {:.0}%)", h, s, l));
            color.insert("hex".to_string(), hex.clone());
            color.insert("rgb".to_string(), format!("rgb({:.0}, {:.0}, {:.0})", h/360.0*255.0, s/100.0*255.0, l/100.0*255.0));
            self.colors.push(color);
        }
    }

    fn print_palette(&self) {
        println!("\n📊 Палитра из {} цветов{}", self.colors.len(), if self.seed != 0 { format!(" (seed: {})", self.seed) } else { "".to_string() });
        println!();

        // Цветные блоки
        for c in &self.colors {
            let hex = &c["hex"][1..];
            let r = u8::from_str_radix(&hex[0..2], 16).unwrap();
            let g = u8::from_str_radix(&hex[2..4], 16).unwrap();
            let b = u8::from_str_radix(&hex[4..6], 16).unwrap();
            print!("\x1b[48;2;{};{};{}m██████████\x1b[0m ", r, g, b);
        }
        println!();
        for c in &self.colors {
            print!("  {:<8}", c["hex"]);
        }
        println!();
    }

    fn save_json(&self, filename: &str) {
        let data = serde_json::json!({
            "seed": self.seed,
            "colors": self.colors
        });
        let json = serde_json::to_string_pretty(&data).unwrap();
        fs::write(filename, json).unwrap();
        println!("💾 Сохранено JSON: {}", filename);
    }

    fn save_html(&self, filename: &str) {
        let mut html = String::from(r#"<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Pastel Palette</title>
<style>
body { font-family: monospace; background: #f5f5f5; padding: 20px; }
.palette { display: flex; gap: 20px; justify-content: center; margin-top: 30px; flex-wrap: wrap; }
.color { width: 120px; height: 150px; border-radius: 12px; display: flex; flex-direction: column; align-items: center; justify-content: flex-end; padding: 10px; color: #333; font-weight: bold; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }
.hex { background: rgba(255,255,255,0.7); padding: 4px 8px; border-radius: 4px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🎨 Pastel Palette</h1>
<div class="palette">
"#);
        for c in &self.colors {
            html.push_str(&format!(r#"<div class="color" style="background: {};">"#, c["hex"]));
            html.push_str(&format!(r#"<span class="hex">{}</span>"#, c["hex"]));
            html.push_str(&format!(r#"<span style="font-size:12px;">{}</span>"#, c["hsl"]));
            html.push_str("</div>");
        }
        html.push_str(r#"
</div>
</body>
</html>"#);
        fs::write(filename, html).unwrap();
        println!("💾 Сохранено HTML: {}", filename);
    }
}

fn hsl_to_rgb(h: f64, s: f64, l: f64) -> (f64, f64, f64) {
    let c = (1.0 - (2.0 * l - 1.0).abs()) * s;
    let x = c * (1.0 - ((h / 60.0) % 2.0 - 1.0).abs());
    let m = l - c / 2.0;
    let (r, g, b) = if h < 60.0 {
        (c, x, 0.0)
    } else if h < 120.0 {
        (x, c, 0.0)
    } else if h < 180.0 {
        (0.0, c, x)
    } else if h < 240.0 {
        (0.0, x, c)
    } else if h < 300.0 {
        (x, 0.0, c)
    } else {
        (c, 0.0, x)
    };
    (r + m, g + m, b + m)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut count = 5;
    let mut seed = 0;
    let mut json_file = "palette.json".to_string();
    let mut html_file = "palette.html".to_string();

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--count" | "-c" => { count = args[i+1].parse().unwrap_or(5); i += 2; }
            "--seed" | "-s" => { seed = args[i+1].parse().unwrap_or(0); i += 2; }
            "--output-json" | "-j" => { json_file = args[i+1].clone(); i += 2; }
            "--output-html" | "-h" => { html_file = args[i+1].clone(); i += 2; }
            _ => { i += 1; }
        }
    }

    println!("🎨 Pastel Palette Generator (Rust)");
    let mut gen = PastelPalette::new(count, seed);
    gen.generate();
    gen.print_palette();
    gen.save_json(&json_file);
    gen.save_html(&html_file);
}
